cache/fast_lru_cache.h

```diff
@@ -105,54 +131,96 @@ struct LRUHandle {
+  // if none is found.
+  LRUHandle* Insert(LRUHandle* h, LRUHandle** old);
+  // Returns a pointer to the handle matching the key/hash, or nullptr
+  // if not present. Assumes the element is in the table.
   LRUHandle* Remove(const Slice& key, uint32_t hash);
 
+  void Exclude(LRUHandle* e);
+
+  void Assign(int slot, LRUHandle* src);
+
   template <typename T>
   void ApplyToEntriesRange(T func, uint32_t index_begin, uint32_t index_end) {
     for (uint32_t i = index_begin; i < index_end; i++) {
-      LRUHandle* h = list_[i];
-      while (h != nullptr) {
-        auto n = h->next_hash;
-        assert(h->InCache());
+      LRUHandle* h = &array_[i];
+      if (h->IsVisible()) {
         func(h);
-        h = n;
       }
     }
   }
 
-  int GetLengthBits() const { return length_bits_; }
+  uint8_t GetLengthBits() const { return length_bits_; }
 
-  // Return the address of the head of the chain in the bucket given
-  // by the hash.
-  inline LRUHandle** Head(uint32_t hash);
+  uint32_t GetOccupancy() const { return occupancy_; }
 
  private:
-  // Return a pointer to slot that points to a cache entry that
-  // matches key/hash.  If there is no such cache entry, return a
-  // pointer to the trailing slot in the corresponding linked list.
-  LRUHandle** FindPointer(const Slice& key, uint32_t hash);
-
-  // Number of hash bits (upper because lower bits used for sharding)
-  // used for table index. Length == 1 << length_bits_
-  int length_bits_;
-
-  // The table consists of an array of buckets where each bucket is
-  // a linked list of cache entries that hash into the bucket.
-  std::unique_ptr<LRUHandle*[]> list_;
+  int FindElement(const Slice& key, uint32_t hash, int& probe,
+                  int displacement);
+
+  int FindPresentElement(const Slice& key, uint32_t hash, int& probe,
+                         int displacement);
+
+  int FindAvailableSlot(const Slice& key, int& probe, int displacement);
+
+  int FindPresentElementOrAvailableSlot(const Slice& key, uint32_t hash,
+                                        int& probe, int displacement);
+
+  // Returns the index of the first slot probed that has a handle e
+  // such that cond(e) is true.  If the probe sequence ends without
+  // a match, returns a pointer to the first slot available in the
+  // sequence.  Otherwise, if no match is found and no slot is
+  // available, returns -1.  For every handle e probed, except the
+  // matching slot, we update e->displacements += displacement.
+  // The variable probe is modified such that consecutive calls to
+  // FindSlot continue probing where the previous left.
+  int FindSlot(const Slice& key, std::function<bool(LRUHandle*)> cond,
+               int& probe, int displacement);
+
+  // Number of hash bits used for table index.
+  // The size of the table is 1 << length_bits_.
+  uint8_t length_bits_;
+
+  // Number of elements in the table.
+  uint32_t occupancy_;
+
+  std::unique_ptr<LRUHandle[]> array_;
```

Comment:
Are there some special reasons to use array here? Can it be std::vector or std::array ?

Reply:
We need a fixed-size array here, so I guess we could use std::array. We wouldn't use any of the container functions, or apply any container algorithms, though. The performance penalty would be minuscule, but still positive. Is it worth it?

Reply:
Can the following be better than std::unique_ptr<LRUHandle[]> ? One concern is that it is not easy to find a continuous memory for a large array_ with element of LRUHandle.
std::unique_ptr<LRUHandle* []>

Reply:
> I guess we could use std::array

That's only for compile-time fixed sizes, which is not what we want here.

> One concern is that it is not easy to find a continuous memory for a large array_

Virtual memory. This is not a concern. The whole point of this change is improving efficiency by consolidating functions of hash table management with memory management, so that we aren't paying for both independently.